src/map.h

```diff
@@ -17,36 +17,32 @@ struct FindPathParams;
 struct AStarNode
 {
 	AStarNode* parent;
-	int_fast32_t f;
+	double g;
+	double f;
 	uint16_t x, y;
 };
 
-static constexpr int32_t MAX_NODES = 512;
-
-static constexpr int32_t MAP_NORMALWALKCOST = 10;
-static constexpr int32_t MAP_DIAGONALWALKCOST = 25;
+static constexpr float MAP_NORMALWALKCOST = 1.0f;
+static constexpr float MAP_DIAGONALWALKCOST = 2.5f;
 
 class AStarNodes
 {
 public:
-	AStarNodes(uint32_t x, uint32_t y);
+	AStarNodes(uint16_t x, uint16_t y);
+
+	void createNewNode(AStarNode* parent, uint16_t x, uint16_t y, double g, double f);
+	void addNode(AStarNode* node) { nodes.push_back(node); };
 
-	AStarNode* createOpenNode(AStarNode* parent, uint32_t x, uint32_t y, int_fast32_t f);
 	AStarNode* getBestNode();
-	void closeNode(AStarNode* node);
-	void openNode(AStarNode* node);
-	int_fast32_t getClosedNodes() const;
-	AStarNode* getNodeByPosition(uint32_t x, uint32_t y);
+	AStarNode* getNodeByPosition(uint16_t x, uint16_t y) { return nodeMap[x][y]; };
 
-	static int_fast32_t getMapWalkCost(AStarNode* node, const Position& neighborPos);
-	static int_fast32_t getTileWalkCost(const Creature& creature, const Tile* tile);
+	static double calculateHeuristic(const Position& p1, const Position& p2);
+	static double getMapWalkCost(AStarNode* node, const Position& neighborPos);
+	static double getTileWalkCost(const Creature& creature, const Tile* tile);
 
 private:
-	AStarNode nodes[MAX_NODES];
-	bool openNodes[MAX_NODES];
-	std::unordered_map<uint32_t, AStarNode*> nodeTable;
-	size_t curNode;
-	int_fast32_t closedNodes;
+	std::vector<AStarNode*> nodes;
+	std::map<uint16_t, std::map<uint16_t, AStarNode*>> nodeMap;
```

Comment:
You never call `clear()` on this `map` either, this could be a potential memory leak, unless it's intended to keep this data static

Reply:
Is it needed? Wont the garbage collector handle that?

Reply:
There is no garbage collector in c++
You could convert those `AStarNode*` into `unique_ptr` this way they would free themselves once they run out of lifetime (reference count) but raw pointers only get deleted when you explicit delete them
```c++
AStarNode* newNode = new AStarNode;
delete newNode // frees the pointer
```